### PR TITLE
Apply StateProof transactions on catchup

### PR DESCRIPTION
**Pull request by Aharonee**

### CODE REVIEW

File: ledger/internal/eval.go [hunk 1]
```diff
@@ -1176,14 +1176,9 @@ func (eval *BlockEvaluator) applyTransaction(tx transactions.Transaction, cow *r
...
 	case protocol.StateProofTx:
-		// in case of a StateProofTx transaction, we want to "apply" it only in validate or generate mode. This will deviate the cow's StateProofNextRound depending on
-		// whether we're in validate/generate mode or not, however - given that this variable is only being used in these modes, it would be safe.
-		// The reason for making this into an exception is that during initialization time, the accounts update is "converting" the recent 320 blocks into deltas to
-		// be stored in memory. These deltas don't care about the state proofs, and so we can improve the node load time. Additionally, it save us from
-		// performing the validation during catchup, which is another performance boost.
-		if eval.validate || eval.generate {
-			err = apply.StateProof(tx.StateProofTxnFields, tx.Header.FirstValid, cow, eval.validate)
-		}
+		// Applying the StateProof transaction will deviate the cow's StateProofNextRound field.
```
Comment (almog-t) [inline ledger/internal/eval.go:1179]: deviate -> modify?

File: ledger/apply/stateproof.go [hunk 2]
```diff
@@ -48,28 +48,25 @@ func StateProof(tx transactions.StateProofTxnFields, atRound basics.Round, sp St
...
+		}
+
 		if err = verify.ValidateStateProof(verificationContext, &tx.StateProof, atRound, &tx.Message); err != nil {
 			return err
 		}
 	}
 
-	sp.SetStateProofNextRound(lastRoundInInterval + basics.Round(config.Consensus[verificationContext.Version].StateProofInterval))
+	// IMPORTANT: this line does not support changing the StateProofInterval consensus param
+	sp.SetStateProofNextRound(lastRoundInInterval + basics.Round(sp.ConsensusParams().StateProofInterval))
```
Comment (almog-t) [inline ledger/apply/stateproof.go:69]: Comment with a bit more detail on why we've decided to use this temporary fix, at the moment it looks strange that we're not using our verificationContext even though it already contains the version.

File: ledger/internal/eval_test.go [hunk 3]
```diff
@@ -382,15 +385,17 @@ func TestCowStateProof(t *testing.T) {
...
+
+	a.Equal(basics.Round(512), c0.mods.StateProofNext)
+	// transaction should be applied without stateproof validation (no context, blockheader or valid stateproof needed as it represents a node catching up)
+	err = apply.StateProof(stateProofTx, atRound, c0, false)
+	a.NoError(err)
+	// make sure that the StateProofNext was updated correctly after applying
+	a.Equal(basics.Round(512+config.Consensus[protocol.ConsensusFuture].StateProofInterval), c0.mods.StateProofNext)
 }
 
 // a couple trivial tests that don't need setup
```
Comment (almog-t) [inline ledger/internal/eval_test.go:401]: Nowhere good to really put this - how about adding (or modifying) a test to check the behavior during reloadLedger at ledger_test.go?

**Review by almog-t — COMMENTED** [review]
lgtm, maybe add a test for the ledger

File: ledger/internal/eval.go (hunk 1 above)
Comment (Aharonee) [reply ledger/internal/eval.go:1179]: I think 'advance' maybe

Comment (almog-t) [reply ledger/internal/eval.go:1179]: go for it

**Review by algoidan — APPROVED** [review]

**Pull request merged**